storage: ensure that the push txn queue notices pushee txn's priority…

spencerkimball opened the pull request:
… upgrade

Previously the pushee txn was only be queried at its expected expiration.
However, this would miss upgrades to its priority since the time the pusher
entered the push txn queue.

Fixes #14670

spencerkimball committed:
storage: ensure that the push txn queue notices pushee txn's priority upgrade

Previously the pushee txn was only be queried at its expected expiration.
However, this would miss upgrades to its priority since the time the pusher
entered the push txn queue.

Fixes #14670

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14789)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSu3t2uUNYDftrD9nb:-KhSu3t2uUNYDftrD9nc:btyx7j4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/base/constants.go#L27)):*
> ```Go
> 	// transaction fails to be heartbeat within 2x the heartbeat interval,
> 	// it may be aborted by conflicting txns.
> 	DefaultHeartbeatInterval = 1 * time.Second
> ```

Why did this setting need to change?

---

*[pkg/storage/push_txn_queue.go, line 289 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSukjhr5clhJmtfqvo:-KhSukjhr5clhJmtfqvp:bhofehq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/storage/push_txn_queue.go#L289)):*
> ```Go
> 	var queryCh <-chan time.Time
> 	var r retry.Retry
> 	//retryOpts := ptq.store.cfg.RangeRetryOptions
> ```

Delete.

---

*[pkg/storage/push_txn_queue.go, line 379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSuJXkZ6RZE6i28MPq:-KhSuJXlPPIL-O5DaYAl:b-xetakr) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/storage/push_txn_queue.go#L379)):*
> ```Go
> 								req.PusherTxn.ID.Short(),
> 								req.PusheeTxn.ID.Short(),
> 								dependencies)
> ```

Did you intend to leave this in? Seems like it could be noisy?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSu3t2uUNYDftrD9nb:-KhSv0Tu1H_nTOGIOulX:b-a9zwvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/base/constants.go#L27)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why did this setting need to change?
</blockquote></details>

This is along the same lines as our overall goal of avoiding long pauses due to cautious timeouts. If a coordinator is interrupted, it would previously leave any blocked transactions waiting for up to 10s. Now it's 2s.

As mentioned in the PR description, txns are going to be unlikely to last longer than a second. For the presumably small percentage that do, one heartbeat per second isn't excessive.

---

*[pkg/storage/push_txn_queue.go, line 289 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSukjhr5clhJmtfqvo:-KhSvYBcFxOdaKUoEAr_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/storage/push_txn_queue.go#L289)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Delete.
</blockquote></details>

Done.

---

*[pkg/storage/push_txn_queue.go, line 379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSuJXkZ6RZE6i28MPq:-KhSvcGbfCZ1JzHyp5TN:bvpooxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/storage/push_txn_queue.go#L379)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you intend to leave this in? Seems like it could be noisy?
</blockquote></details>

No, I put it behind a `log.V(1)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though it is probably worthwhile to have @bdarnell look at this as he did a more thorough review of `PushTxnQueue`.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/base/constants.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSu3t2uUNYDftrD9nb:-KhSwJZZCWTb-ejCNDC_:bpylb5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/base/constants.go#L27)):*
> As mentioned in the PR description, txns are going to be unlikely to last longer than a second. For the presumably small percentage that do, one heartbeat per second isn't excessive.

Ok, with the reasoning here, but I'm not seeing it mentioned in the PR/commit message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-:-KhSw_cKehRnXFO32EA7:b45fbjg)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhSwkjjHnyvE5aYaxLr:b-6o262n) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
> ```Go
> 					_, haveDependency := push.mu.dependents[*req.PusheeTxn.ID]
> 					var dependencies string
> 					if haveDependency || log.V(2) {
> ```

shouldn't this be `|| log.V(2)` to match the condition where `dependencies` is referenced below?

This code seems a bit confused at the moment.

---

*[pkg/storage/push_txn_queue.go, line 356 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwcK3z5m-nyhLMKrq:-KhSwcK3z5m-nyhLMKrr:b-bn5u10) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L356)):*
> ```Go
> 					var dependencies string
> 					if haveDependency || log.V(2) {
> 						var dependents []string
> ```

preallocate this?

---

*[pkg/storage/push_txn_queue.go, line 362 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwVDvFPyUP1kywUsZ:-KhSwVDvFPyUP1kywUs_:b-bsdibt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L362)):*
> ```Go
> 						dependencies = strings.Join(dependents, ", ")
> 						if log.V(2) {
> 							log.Infof(ctx, "%s has dependencies=%s", req.PusherTxn.ID.Short(), dependencies)
> ```

i think you can just log `dependents` here; it will show up as `[foo, bar, baz]`

---

*[pkg/storage/push_txn_queue.go, line 377 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwOTV7gvJv5yF3VtN:-KhSwOTWJgWIGXfqO93P:b-b0ufuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L377)):*
> ```Go
> 									req.PusherTxn.ID.Short(),
> 									req.PusheeTxn.ID.Short(),
> 									dependencies)
> ```

nit: closing paren on next line for symmetry.

---

*[pkg/storage/store.go, line 2614 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSOPWTMItz8hkMLN_U:-KhSOPWTMItz8hkMLN_V:b-2lw5v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/storage/store.go#L2614)):*
> ```Go
> 			}
> 			// Move the push timestamp forward to the current time, as this
> 			// request may have been waiting to pushe the txn. If we don't
> ```

s/pushe/push/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSu3t2uUNYDftrD9nb:-KhT26SPUC4PzaunJSUs:b-ny0hlq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/base/constants.go#L27)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

> As mentioned in the PR description, txns are going to be unlikely to last longer than a second. For the presumably small percentage that do, one heartbeat per second isn't excessive.

Ok, with the reasoning here, but I'm not seeing it mentioned in the PR/commit message.
</blockquote></details>

This makes me more nervous than reducing the lease timeout. There, we know the impact is more or less bounded by the number of nodes. Here, the number of additional heartbeats scales with the number of transactions (albeit at a much lower constant factor since most transactions are sub-second). This has more potential for escalating a performance anomaly into a larger outage.

---

*[pkg/storage/push_txn_queue.go, line 288 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSzsWc0FhC02INaAVw:-KhSzsWc0FhC02INaAVx:bmbsrqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1357da1b07737fe11f7730e8a63290f8ba8407/pkg/storage/push_txn_queue.go#L288)):*
> ```Go
> 	var r retry.Retry
> 	if req.PusherTxn.ID != nil {
> 		retryOpts := ptq.store.cfg.RangeRetryOptions
> ```

I think this was the only place `cfg.RangeRetryOptions` was used, so maybe we should just change the defaults for that (or maybe get rid of it and hard-code things here but it seems potentially useful to leave the configuration scaffolding in place)

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhT-YsTdgBSVL_4OajX:b-4i7j76) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

shouldn't this be `|| log.V(2)` to match the condition where `dependencies` is referenced below?

This code seems a bit confused at the moment.
</blockquote></details>

I think it should be `log.V(1)` since we end up using `dependencies` on both `V(1)` and `V(2)` paths.  And shouldn't it be `&&` instead of `||`? Or even just `if log.V(1)` since `haveDependency ||` seems like an insignificant optimization. 

---

*[pkg/storage/replica_command.go, line 1564 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhT0DkxnMtqOsfc7674:-KhT0DkxnMtqOsfc7675:brm4em9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/replica_command.go#L1564)):*
> ```Go
> 
> 	switch {
> 	case args.Force:
> ```

This is a downstream-of-raft changes that is unsafe to make without propEvalKV. 

---

*[pkg/storage/store.go, line 2691 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhT18W77JrGRvyE4Zap:-KhT18W77JrGRvyE4Zaq:b8zp2s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/store.go#L2691)):*
> ```Go
> 				if ba.Txn != nil && ba.Txn.LastHeartbeat == nil {
> 					now := s.Clock().Now()
> 					ba.Txn.LastHeartbeat = &now
> ```

Lying about `LastHeartbeat` makes me nervous. Advancing `LastHeartbeat` when no heartbeat has taken place may cause a [new txn record to be created for a previously-aborted txn](https://github.com/cockroachdb/cockroach/blob/a39be9a240c12a09d35d2984fe92a9c757a69b0d/pkg/storage/replica_command.go#L517-L525)

Where exactly is the transaction "expiring on arrival"? Can we recover from this case by doing a QueryTxn to get the latest update (assuming the coordinator has in fact been heartbeating)? Do we need to preemptively QueryTxn to get the current state without trying and waiting for a failure?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/base/constants.go, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSu3t2uUNYDftrD9nb:-KhT4KUee9zT-9Bogutk:b-8bh5iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/base/constants.go#L27)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This makes me more nervous than reducing the lease timeout. There, we know the impact is more or less bounded by the number of nodes. Here, the number of additional heartbeats scales with the number of transactions (albeit at a much lower constant factor since most transactions are sub-second). This has more potential for escalating a performance anomaly into a larger outage.
</blockquote></details>

Another way to look at this in terms of getting comfortable with it is to do some basic reasoning about order of extra RPCs. As you say, this is tied to the number of txns being sent through the system. Because txns are synchronous from most clients' perspectives, it's reasonable to consider total txns in the system as proportional to work being done, regardless of bugs which might cause them to stall and remain active for long enough to start heartbeating.

With that in mind, it seems reasonable to consider that an extant transaction which is being heartbeat once per second is likely to be generating less load than a succession of fast txns which never get to the point of requiring heartbeats.

So I think we're less likely to escalate into a performance anomaly due to a txn heartbeat storm than simply a throughput slowdown due to hanging txns which are sending a heartbeat per second instead of 10s or 100s of individual txns with fast turnaround.

---

*[pkg/storage/push_txn_queue.go, line 288 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSzsWc0FhC02INaAVw:-KhT7wg-XGTmuH_hF84U:b-i3f9qo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1357da1b07737fe11f7730e8a63290f8ba8407/pkg/storage/push_txn_queue.go#L288)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think this was the only place `cfg.RangeRetryOptions` was used, so maybe we should just change the defaults for that (or maybe get rid of it and hard-code things here but it seems potentially useful to leave the configuration scaffolding in place)
</blockquote></details>

Done. Good point, we'll need them for a unittest I haven't written yet as well.

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhT93DSCJ5Y4lJWX0oe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think it should be `log.V(1)` since we end up using `dependencies` on both `V(1)` and `V(2)` paths.  And shouldn't it be `&&` instead of `||`? Or even just `if log.V(1)` since `haveDependency ||` seems like an insignificant optimization. 
</blockquote></details>

Done.

---

*[pkg/storage/push_txn_queue.go, line 356 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwcK3z5m-nyhLMKrq:-KhT92zUDQyfuJV7R44H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L356)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

preallocate this?
</blockquote></details>

Done.

---

*[pkg/storage/push_txn_queue.go, line 362 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwVDvFPyUP1kywUsZ:-KhT9HJY4CNidMBByOzL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L362)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think you can just log `dependents` here; it will show up as `[foo, bar, baz]`
</blockquote></details>

Done.

---

*[pkg/storage/push_txn_queue.go, line 377 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwOTV7gvJv5yF3VtN:-KhT9JNJ0ouXJBTsZrHS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L377)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: closing paren on next line for symmetry.
</blockquote></details>

Done.

---

*[pkg/storage/replica_command.go, line 1564 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhT0DkxnMtqOsfc7674:-KhT9LaFBnHtiqmryB5Q:biu0ij1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/replica_command.go#L1564)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is a downstream-of-raft changes that is unsafe to make without propEvalKV. 
</blockquote></details>

Ugh, good catch. I've moved it back and added a TODO. It's really just about the reason we're showing for why we abandoned something.

---

*[pkg/storage/store.go, line 2614 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSOPWTMItz8hkMLN_U:-KhT9e3SkQpjDVQ7dWRK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0caf709ea96dceb182a903b5d2994cb4e02c60ec/pkg/storage/store.go#L2614)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/pushe/push/
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 2691 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhT18W77JrGRvyE4Zap:-KhTOSepmaakCEWa05M1:b-i7a2rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/store.go#L2691)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Lying about `LastHeartbeat` makes me nervous. Advancing `LastHeartbeat` when no heartbeat has taken place may cause a [new txn record to be created for a previously-aborted txn](https://github.com/cockroachdb/cockroach/blob/a39be9a240c12a09d35d2984fe92a9c757a69b0d/pkg/storage/replica_command.go#L517-L525)

Where exactly is the transaction "expiring on arrival"? Can we recover from this case by doing a QueryTxn to get the latest update (assuming the coordinator has in fact been heartbeating)? Do we need to preemptively QueryTxn to get the current state without trying and waiting for a failure?
</blockquote></details>

Actually, I think it's just saner to remove this code. It won't make a difference to performance, just potentially the odd unnecessary restart as another waiting txn is able to trivially push this one because it's past its due date after waiting too long.

This problem showed up consistently when the primary bug being fixed here (not updating pushee priority) was still present.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhTVzrsel9iKqet2y28:bw5ul0e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

should the allocation and population of `dependents` be behind `log.V(1)`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhTWZjlrigc3OB6SN3h:b-16v70g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should the allocation and population of `dependents` be behind `log.V(1)`?
</blockquote></details>

It's used below in the deadlock warning message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhTYkzIkOvI2hFq0CFv:b-6sdpvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

It's used below in the deadlock warning message.
</blockquote></details>

and that deadlock message is behind `log.V(1)`, isn't it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/push_txn_queue.go, line 355 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-KhSwkjjHnyvE5aYaxLq:-KhTcwWnHCsqEbMYLoBy:b5fm02u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b304c4ca4194478199406a1890fc18f5971727d/pkg/storage/push_txn_queue.go#L355)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

and that deadlock message is behind `log.V(1)`, isn't it?
</blockquote></details>

Ah. That seems fragile. I'll leave as is. This isn't a normal-case execution path.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14789#-:-KhU7364K3JfqJock1ma:bnfp4nl)*
<!-- Sent from Reviewable.io -->
